File: 2024/01/haskell/01.hs
{- stack script --resolver lts-14.20 -}
module Main where

import Data.List

parse :: [[Int]] -> ([Int], [Int])
parse [] = ([], [])
parse (x:xs) = (p : a, q : b)
  where
    (a, b) = parse xs
    [p, q] = x

task1 :: ([Int], [Int]) -> Int
task1 (a, b) = sum $ zipWith (\x y -> abs $ x - y) a' b'
  where
    a' = sort a
    b' = sort b

task2 :: ([Int], [Int]) -> Int
task2 (a, b) = task2' a' b'
  where
    a' = sort a
    b' = sort b

    task2' :: [Int] -> [Int] -> Int
    task2' []  _ = 0
    task2'  _ [] = 0
    task2'  a  b | ha == hb  = cnta * cntb * ha + task2' a' b'
                 | ha  < hb  = task2' a' b
                 | otherwise = task2' a  b'
      where
        ha   = head a
        hb   = head b
        a'   = dropWhile (== ha) a
        b'   = dropWhile (== hb) b
        cnta = length $ takeWhile (== ha) a
        cntb = length $ takeWhile (== hb) b

main :: IO ()
main = do
  input <- fmap (parse . map (map read . words) . lines) getContents
  putStr "Task 1: "
  print $ task1 input
  putStr "Task 2: "
  print $ task2 input

File: 2024/01/rust/01.rs
use std::io;
use std::iter::zip;
use std::collections::BTreeMap;

fn task1(mut a: Vec<i32>, mut b: Vec<i32>) -> i32 {
    a.sort();
    b.sort();
    zip(a.iter(), b.iter()).map(|(x, y)| (x - y).abs()).sum::<i32>()
}

fn task2(a: Vec<i32>, b: Vec<i32>) -> i32 {
    let mut freq = BTreeMap::new();
    for y in b.iter() {
        *freq.entry(y).or_insert(0) += 1;
    }
    a.iter().map(|x| x * freq.remove(&x).unwrap_or(0)).sum::<i32>()
}

fn main() {
    let mut a = Vec::new();
    let mut b = Vec::new();

    let mut input = String::new();
    while io::stdin().read_line(&mut input).unwrap() > 0 {
        let mut iter = input.split_whitespace();
        let x: i32 = iter.next().unwrap().parse().unwrap();
        let y: i32 = iter.next().unwrap().parse().unwrap();
        a.push(x);
        b.push(y);
        input.clear();
    }

    println!("Task 1: {:?}", task1(a.clone(), b.clone()));
    println!("Task 2: {:?}", task2(a.clone(), b.clone()));
}
